Add unloadability howto document

janvorli opened the pull request:
## Summary

This document describes how to use unloadability in .NET Core 3.0 and
how to debug issues with unloading.

It also contains a link to a complete sample in dotnet/samples repo.

janvorli committed:
Add unloadability howto document

This document describes how to use unloadability in .NET Core 3.0 and
how to debug issues with unloading.

It also contains a link to a complete sample in dotnet/samples repo.

janvorli commented:
cc: @jkotas, @jkoritzinsky, @AaronRobinsonMSFT

jkotas commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+```
+As you can see, the `Load` method returns `null`. That means that all the dependency assemblies are loaded into the default context and only the assemblies explicitly loaded into the new context are in this context. 
+
+In case you want to load some or all of the dependencies into the `AssemblyLoadContext` too, you can use the `AssemblyDependencyResolver` in the `Load` method. The `AssemblyDependencyResolver` resolves the assembly names to absolute assembly file paths using the *.deps.json file contained in the directory of the main assembly loaded into the context and using assembly files in that directory.
````
```suggestion
In case you want to load some or all of the dependencies into the `AssemblyLoadContext` too, you can use the `AssemblyDependencyResolver` in the `Load` method. The `AssemblyDependencyResolver` resolves the assembly names to absolute assembly file paths using the `*.deps.json` file contained in the directory of the main assembly loaded into the context and using assembly files in that directory.
```

AaronRobinsonMSFT replied:
@jkoritzinsky added a plugin sample for this. Reference it?

janvorli replied:
Do you mean a sample that uses *.deps.json in some interesting way or a just a sample that uses AssemblyDependencyResolver? If it is the latter, I already have a link above to my own unloadability sample that uses it.

janvorli replied:
My sample is also in that official sample repo. I've merged it in two days ago. It covers unloadability and since I wanted to make it a boiler plate that people can use, I've used the AssemblyDependencyResolver there too. If you've missed the link in the text above, it is here:
https://github.com/dotnet/samples/tree/master/core/tutorials/Unloading

jkoritzinsky replied:
In that case you should use a code reference tag to automatically import the code from the sample like the following

`[!code-csharp[Assembly-load-context-with-assembly-dependency-resolver](~/samples/blob/master/core/tutorials/Unloading/Host/Program.cs)]`

You might also want to seperate out the ALC into a different file to make the code reference work a little cleaner.

janvorli replied:
I didn't want to include the code from my sample here. I see this doc and the sample as separate entities. This doc focuses on a simple version where the Load method always returns NULL and just mentions the more advanced way briefly so that people know it exist. 
The sample is a separate full blown end to end example of using unloadable ALC that complements this article. I didn't actually know you were working on your sample and frankly even didn't know about the AssemblyDependencyResolver existence when I've completed the first version of my sample and I was originally using just the simple ALC returning NULL from the Load overried. @jkotas then pointed out during review of that sample that it might be nice to make it more complete by adding the usage of the AssemblyDependencyResolver. Thus I've added it there and mentioned it in this doc.

jkoritzinsky replied:
That tag is a specially handled by the markdown engine that docs.microsoft.com uses. Github doesn't understand it fully.

AaronRobinsonMSFT commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+WeakReference testAlcWeakRef;
+int result = ExecuteAndUnload("absolute/path/to/your/assembly", out testAlcWeakRef);
+```
+However, the unloading doesn't complete immediately. As I've already mentioned, it relies on GC to collect all the objects from the test assembly (etc.). In many cases, it is not necessary to wait for the unload completion. However there are cases where it is useful to know that the unload has finished. For example, you may want to delete the assembly file that was loaded into the custom context  ...
````
Please remove all uses of "I". In this case, ~~As I've already mentioned,~~ -> "As previously mentioned,"

AaronRobinsonMSFT commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+}
+```
+### The Unloading event
+In some cases, it may be necessary for the code loaded into a custom AssemblyLoadContext to perform some cleanup when the unloading is initiated. For example, it may need to stop threads, cleanup some strong GC handles, etc. The `Unloading` event can be used in such cases. A handler that performs the necessary cleanup can be hooked to this event.
````
Unloading should be quoted not back-ticked.

AaronRobinsonMSFT commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,359 @@
+  * Type from such an assembly
+  * Instance of a type  from such an assembly
+* Threads running code from an assembly loaded into the custom AssemblyLoadContext
+* Instances of non-collectible AssemblyLoadContext
```
We are inconsistent on back-ticks around `AssemblyLoadContext` and types in general.

AaronRobinsonMSFT commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,359 @@
+* Threads running code from an assembly loaded into the custom AssemblyLoadContext
+* Instances of non-collectible AssemblyLoadContext
+* Pending RegisteredWaitHandle instances
+The above mentioned reference can come from any object with root in a stack slot or processor register (method locals, either explicitly created by the user code or implicitly by the JIT), a static variable or strong / pinning GC handle. TODO: explain what root means
```
>TODO: explain what roo means

Is there a nice definition in the BOTR we can reference? If not, I would add one there.

AaronRobinsonMSFT commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+In LLDB:
+plugin load /path/to/libsosplugin.so
+```
+Let's try to debug an example program that has problems with unloading. I have included its source code below. When you run it under WinDbg, the program breaks into the debugger right after attempting to check for the unload success. We can then start looking for the culprits.
````
~~I have included its source code below~~ -> "Source code is included below."

janvorli commented:
@AaronRobinsonMSFT I've updated the doc based on your feedback. The only place where I've left the type names without backticks is in the section titles, as it looks awkward in those.

janvorli commented:
Oh, I've forgotten about the TODO about explaining gc root. I am going to add that. I have not found it in any doc.

janvorli commented:
I've actually reworded the text that was mentioning the roots before and got rid of the "roots" name there as it doesn't seem to be necessary for the explanation.

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+
+Found 3 roots.
+```
+Now you need to figure out where is the root located so that you can fix it. The easiest case is when the root is a stack slot or a processor register. In that case, the `gcroot` shows you the name of the function whose frame contains the root and the thread executing that function. The difficult case is when the root is a static variable or a GC handle. 
````
```suggestion
Now you need to figure out where the root is located so that you can fix it. The easiest case is when the root is a stack slot or a processor register. In that case, the `gcroot` shows you the name of the function whose frame contains the root and the thread executing that function. The difficult case is when the root is a static variable or a GC handle. 
```

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,359 @@
+
+In our example above, the first root is a local of type `System.Reflection.RuntimeMethodInfo` stored in the frame of the function `example.Program.Main(System.String[])` at address `rbp-20` (`rbp` is the processor register `rbp` and -20 is a hexadecimal offset from that register).
+
+The second root is a normal (strong) `GCHandle` that holds reference to an instance of the `test.Test` class. 
```
```suggestion
The second root is a normal (strong) `GCHandle` that holds a reference to an instance of the `test.Test` class. 
```

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,359 @@
+
+The third root is a pinned `GCHandle`. This one is actually a static variable. Unfortunately, there is no way to tell. Statics for reference types are stored in a managed object array in internal runtime structures.
+
+Another case that can prevent unloading of an `AssemblyLoadContext` is when a thread has a frame of a method from assembly loaded into the `AssemblyLoadContext` on its stack. You can check that by dumping managed call stacks of all threads:
```
```suggestion
Another case that can prevent unloading of an `AssemblyLoadContext` is when a thread has a frame of a method from an assembly loaded into the `AssemblyLoadContext` on its stack. You can check that by dumping managed call stacks of all threads:
```

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+The third root is a pinned `GCHandle`. This one is actually a static variable. Unfortunately, there is no way to tell. Statics for reference types are stored in a managed object array in internal runtime structures.
+
+Another case that can prevent unloading of an `AssemblyLoadContext` is when a thread has a frame of a method from assembly loaded into the `AssemblyLoadContext` on its stack. You can check that by dumping managed call stacks of all threads:
+```
````
```suggestion
```console
```

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+```
+~*e !clrstack
+```
+The command means "apply to all threads the `!clrstack` command". Here is an example of an output of that command for our example. Unfortunately, LLDB on Unix doesn't have any way to apply a command to all threads, so you'll need to resort to manual switching threads and repeating the `clrstack` command.
````
```suggestion
The command means "apply to all threads the `!clrstack` command". Here is the output of that command for our example. Unfortunately, LLDB on Unix doesn't have any way to apply a command to all threads, so you'll need to resort to manually switching threads and repeating the `clrstack` command.
```

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+```
+The command means "apply to all threads the `!clrstack` command". Here is an example of an output of that command for our example. Unfortunately, LLDB on Unix doesn't have any way to apply a command to all threads, so you'll need to resort to manual switching threads and repeating the `clrstack` command.
+You should ignore all threads where the debugger says "Unable to walk the managed stack."
+```
````
```suggestion
```console
```

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+0000001fc727f7f0 00007ffae5fdc1e3 [DebuggerU2MCatchHandlerFrame: 0000001fc727f7f0] 
+
+```
+As you can see the last thread has `test.Program.ThreadProc()`. This is a function from the assembly loaded into the `AssemblyLoadContext` and so it keeps the `AssemblyLoadContext` alive.
````
```suggestion
As you can see, the last thread has `test.Program.ThreadProc()`. This is a function from the assembly loaded into the `AssemblyLoadContext`, and so it keeps the `AssemblyLoadContext` alive.
```

rpetrusha commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,359 @@
+## Example source with unloadability issues
+This example is used in the debugging above.
+### Main testing program
+```C#
````
```suggestion
```csharp
```
But here again, it's best to put this in the samples repo.

rpetrusha reviewed (changes requested):
This looks good, @janvorli. I've left a number of questions, comments, and suggestions for you to consider. You can bulk-review my suggested changes by reviewing from the **Files changes** tab.

janvorli commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,359 @@
+
+There is one noteworthy difference between the unloading using `AssemblyLoadContext` and using AppDomains. With AppDomains, the unloading is forced. At the unload time, all threads running in the target AppDomain are aborted, managed COM objects created in the AppDomain that's being unloaded are destroyed, etc. 
+
+With the new way that uses `AssemblyLoadContext`, the unload is "cooperative". Calling the `Unload` method on the `AssemblyLoadContext` just initiates the unloading. But the unload will not complete until there are no threads having code from the assemblies loaded into the `AssemblyLoadContext` on their call stacks and until there are no strong references to types from the assemblies loaded into  ...
```
"their instances" means instances of the types from the assemblies  ... I can see it is a bit unclear, I'll reword it.

janvorli commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,359 @@
+There is one noteworthy difference between the unloading using `AssemblyLoadContext` and using AppDomains. With AppDomains, the unloading is forced. At the unload time, all threads running in the target AppDomain are aborted, managed COM objects created in the AppDomain that's being unloaded are destroyed, etc. 
+
+With the new way that uses `AssemblyLoadContext`, the unload is "cooperative". Calling the `Unload` method on the `AssemblyLoadContext` just initiates the unloading. But the unload will not complete until there are no threads having code from the assemblies loaded into the `AssemblyLoadContext` on their call stacks and until there are no strong references to types from the assemblies loaded into  ...
+## Using unloadable AssemblyLoadContext
```
Collectible and unloadable AssemblyLoadContext means the same thing. So it is an AssemblyLoadContext created with the isCollectible argument of the constructor set to true. While the underlying features that it uses are collectible Assemblies (which means they can be collected by the GC after the AssemblyLoadContext is unloaded / collected), the term "unloadability" is what people use in connection to the AssemblyLoadContext as in the desktop .NET framework, the equivalent functionality was called AppDomain unloading (it doesn't use collectibility).

janvorli replied:
After giving it more thought, I've changed all references to "unloadable AssemblyLoadContext" to "collectible AssemblyLoadContext". It seems it is really better that way.

janvorli commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,359 @@
+There is one noteworthy difference between the unloading using `AssemblyLoadContext` and using AppDomains. With AppDomains, the unloading is forced. At the unload time, all threads running in the target AppDomain are aborted, managed COM objects created in the AppDomain that's being unloaded are destroyed, etc. 
+
+With the new way that uses `AssemblyLoadContext`, the unload is "cooperative". Calling the `Unload` method on the `AssemblyLoadContext` just initiates the unloading. But the unload will not complete until there are no threads having code from the assemblies loaded into the `AssemblyLoadContext` on their call stacks and until there are no strong references to types from the assemblies loaded into  ...
+## Using unloadable AssemblyLoadContext
```
I am not sure I understand the suggested change, it seems to change the meaning of it. It means "How to use unloadable AssemblyLoadContext", but in your wording, it sounds like an imperative.

janvorli commented:
@rpetrusha thank you so much for the detailed review and all the suggested corrections of the grammar. I can see I should get better especially in the comma usage.
I've applied all of your suggestions except for changing one of the titles and also changed "unloadable AssemblyLoadContext" at few places to "collectible AssemblyLoadContext".
I have also created a PR containing all the code snippets from this document here: https://github.com/dotnet/samples/pull/617. I will update this doc once that PR goes in to reference the snippets instead of having them inline.

AaronRobinsonMSFT commented:
LGTM.

mairaw commented:
There's a broken link in the build report and the article needs to be added to the TOC file as well (https://github.com/dotnet/docs/blob/master/docs/toc.md)

rpetrusha commented:
The broken link in the build report is to a .NET Core 3.0 API, @mairaw. I thought that it would be better to leave the xref than fence it and have to change it later.

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+### Use a custom collectible AssemblyLoadContext
+This section assumes the simpler version of the `TestAssemblyLoadContext` is being used.
+
+You can create an instance of the custom `AssemblyLoadContext` and load an assembly into it as follows:
```
```suggestion
You can create an instance of the custom `AssemblyLoadContext` and load an assembly into it as follows:

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```csharp
+var alc = new TestAssemblyLoadContext();
+Assembly a = alc.LoadFromAssemblyPath("absolute/path/to/your/assembly");
+```
````
```suggestion
```

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```
+For each of the assemblies referenced by the loaded assembly, the `TestAssemblyLoadContext.Load` method is called so that the `TestAssemblyLoadContext` can decide where to get the assembly from. In our case, it returns `null` to indicate that it should be loaded into the default context from locations that the runtime uses to load assemblies by default.
+
+Now that we have loaded an assembly, we can execute a method from it. Let's run the `Main` method:
````
We don't use we in our content, unless it's a Microsoft recommendation. Suggestion:

```suggestion
Now that an assembly was loaded, you can execute a method from it. Run the `Main` method:
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+var args = new object[1] {new string[] {"Hello"}};
+int result = (int)a.EntryPoint.Invoke(null, args);
+```
+After the `Main` method returns, we can initiate unloading by either calling the `Unload` method on the custom `AssemblyLoadContext` or getting rid of the reference we have to the `AssemblyLoadContext`:
````
```suggestion

After the `Main` method returns, you can initiate unloading by either calling the `Unload` method on the custom `AssemblyLoadContext` or getting rid of the reference you have to the `AssemblyLoadContext`:
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```csharp
+alc.Unload();
+```
+This is sufficient to unload the test assembly. Let's actually put all of this into a separate non-inlineable method to ensure that the `TestAssemblyLoadContext`, `Assembly` and `MethodInfo` (the `Assembly.EntryPoint`) can't be kept alive by stack slot references (real- or JIT-introduced locals). That could keep the `TestAssemblyLoadContext` alive and prevent the unload.
````
```suggestion

This is sufficient to unload the test assembly. Let's actually put all of this into a separate non-inlineable method to ensure that the `TestAssemblyLoadContext`, `Assembly`, and `MethodInfo` (the `Assembly.EntryPoint`) can't be kept alive by stack slot references (real- or JIT-introduced locals). That could keep the `TestAssemblyLoadContext` alive and prevent the unload.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```
+This is sufficient to unload the test assembly. Let's actually put all of this into a separate non-inlineable method to ensure that the `TestAssemblyLoadContext`, `Assembly` and `MethodInfo` (the `Assembly.EntryPoint`) can't be kept alive by stack slot references (real- or JIT-introduced locals). That could keep the `TestAssemblyLoadContext` alive and prevent the unload.
+
+Let's also return a weak reference to the `AssemblyLoadContext` so that we can use it later to detect unload completion.
````
```suggestion
Also, return a weak reference to the `AssemblyLoadContext` so that you can use it later to detect unload completion.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+    return result;
+}
+```
+Now we can run this function to load, execute, and unload the assembly.
````
```suggestion

Now you can run this function to load, execute, and unload the assembly.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+WeakReference testAlcWeakRef;
+int result = ExecuteAndUnload("absolute/path/to/your/assembly", out testAlcWeakRef);
+```
+However, the unload doesn't complete immediately. As previously mentioned, it relies on the GC to collect all the objects from the test assembly. In many cases, it is not necessary to wait for the unload completion. However there are cases where it is useful to know that the unload has finished. For example, you may want to delete the assembly file that was loaded into the custom `AssemblyLoadCon ...
````
```suggestion

However, the unload doesn't complete immediately. As previously mentioned, it relies on the GC to collect all the objects from the test assembly. In many cases, it isn't necessary to wait for the unload completion. However, there are cases where it's useful to know that the unload has finished. For example, you may want to delete the assembly file that was loaded into the custom `AssemblyLoadContext` from disk. In such a case, the following code snippet can be used. It triggers a GC and waits for pending finalizers in a loop until the weak reference to the custom `AssemblyLoadContext` is set to `null`, indicating the target object was collected. Note that in most cases, just one pass through the loop is required. However, for more complex cases where objects created by the code running in the `AssemblyLoadContext` have finalizers, more passes may be needed.

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+    GC.WaitForPendingFinalizers();
+}
+```
+### The Unloading event
````
```suggestion

### The Unloading event

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```
+### The Unloading event
+In some cases, it may be necessary for the code loaded into a custom `AssemblyLoadContext` to perform some cleanup when the unloading is initiated. For example, it may need to stop threads, clean up some strong GC handles, etc. The `Unloading` event can be used in such cases. A handler that performs the necessary cleanup can be hooked to this event.
+### Troubleshooting unloadability issues
````
```suggestion

### Troubleshoot unloadability issues

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+### The Unloading event
+In some cases, it may be necessary for the code loaded into a custom `AssemblyLoadContext` to perform some cleanup when the unloading is initiated. For example, it may need to stop threads, clean up some strong GC handles, etc. The `Unloading` event can be used in such cases. A handler that performs the necessary cleanup can be hooked to this event.
+### Troubleshooting unloadability issues
+Due to the cooperative nature of the unloading, it is easy to forget about references keeping the stuff in a collectible `AssemblyLoadContext` alive and preventing unload. Here is a summary of things (some of them non-obvious) that can hold the references:
```
```suggestion
Due to the cooperative nature of the unloading, it's easy to forget about references keeping the stuff in a collectible `AssemblyLoadContext` alive and preventing unload. Here is a summary of things (some of them non-obvious) that can hold the references:

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+* Instances of custom non-collectible `AssemblyLoadContext` types created inside of the collectible `AssemblyLoadContext`
+* Pending <xref:System.Threading.RegisteredWaitHandle> instances with callbacks set to methods in the custom `AssemblyLoadContext`
+
+Hints to find stack slot / processor register rooting an object:
```
```suggestion
Hints to find stack slot / processor register rooting an object:

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+The .NET Core 3.0 supports unloadability through <xref:System.Runtime.Loader.AssemblyLoadContext>. You can load a set of assemblies into a collectible `AssemblyLoadContext`, execute methods in them or just inspect them using reflection, and finally unload the `AssemblyLoadContext`. That unloads the assemblies loaded into that `AssemblyLoadContext`.
+
+There is one noteworthy difference between the unloading using `AssemblyLoadContext` and using AppDomains. With AppDomains, the unloading is forced. At the unload time, all threads running in the target AppDomain are aborted, managed COM objects created in the target AppDomain are destroyed, etc. With `AssemblyLoadContext`, the unload is "cooperative". Calling the <xref:System.Runtime.Loader.Asse ...
+- No threads have code from the assemblies loaded into the `AssemblyLoadContext` on their call stacks.
```
Following the sentence as a whole reads a bit weird...

It doesn't complete until no threads have code from the assemblies loaded into the `AssemblyLoadContext` on their call stacks.

janvorli replied:
I've reworded this part to make it easier to understand.

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+# Using and debugging unloadability in .NET Core
+An ability to load and later unload a set of assemblies is one of the features that was missing until now in .NET Core. In .NET Framework, custom app domains are used for this purpose. But since .NET Core supports only a single default app domain, how can we achieve that goal in .NET Core?
```
These first few paragraphs read a lot like a blog post to me. Also, we need to think of topics as future proof. So saying things like now, etc. should be avoided.

A suggestion on how to get this started:

```suggestion
Starting with .NET Core 3.0, the ability to load and later unload a set of assemblies is supported. In .NET Framework, custom app domains were used for this purpose, but .NET Core only supports a single default app domain.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+# Using and debugging unloadability in .NET Core
+An ability to load and later unload a set of assemblies is one of the features that was missing until now in .NET Core. In .NET Framework, custom app domains are used for this purpose. But since .NET Core supports only a single default app domain, how can we achieve that goal in .NET Core?
+
+The .NET Core 3.0 supports unloadability through <xref:System.Runtime.Loader.AssemblyLoadContext>. You can load a set of assemblies into a collectible `AssemblyLoadContext`, execute methods in them or just inspect them using reflection, and finally unload the `AssemblyLoadContext`. That unloads the assemblies loaded into that `AssemblyLoadContext`.
```
```suggestion
.NET Core 3.0 and later versions support unloadability through <xref:System.Runtime.Loader.AssemblyLoadContext>. You can load a set of assemblies into a collectible `AssemblyLoadContext`, execute methods in them or just inspect them using reflection, and finally unload the `AssemblyLoadContext`. That unloads the assemblies loaded into that `AssemblyLoadContext`.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+* If a reference to an object was available at any point in a method, the JIT might have decided to keep the reference in a stack slot / processor register for as long as it wants in the current function.
+## Debugging unloading issues
+Debugging issues with unloading can be tedious. You can get into situations where you don't know what can be holding an `AssemblyLoadContext` alive, but the unload fails.
+The best weapon to help with that is Windbg (LLDB on Unix) with the SOS plugin. We need to find what's keeping a `LoaderAllocator` belonging to the specific `AssemblyLoadContext` alive.
```
```suggestion
The best weapon to help with that is Windbg (LLDB on Unix) with the SOS plugin. You need to find what's keeping a `LoaderAllocator` belonging to the specific `AssemblyLoadContext` alive.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```console
+plugin load /path/to/libsosplugin.so
+```
+Let's try to debug an example program that has problems with unloading. Source code is included below. When you run it under WinDbg, the program breaks into the debugger right after attempting to check for the unload success. We can then start looking for the culprits.
````
```suggestion
Let's try to debug an example program that has problems with unloading. Source code is included below. When you run it under WinDbg, the program breaks into the debugger right after attempting to check for the unload success. You can then start looking for the culprits.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```
+Let's try to debug an example program that has problems with unloading. Source code is included below. When you run it under WinDbg, the program breaks into the debugger right after attempting to check for the unload success. We can then start looking for the culprits.
+
+Note that if you debug using LLDB on Unix, the SOS commands in the examples below don't have the `!` in front of them.
````
don't use below

```suggestion
Note that if you debug using LLDB on Unix, the SOS commands in the following examples don't have the `!` in front of them.

```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```console
+!dumpheap -type LoaderAllocator
+```
+It will dump all objects with a type name containing `LoaderAllocator` that are in the GC heap. Here is an example:
````
```suggestion
This command dumps all objects with a type name containing `LoaderAllocator` that are in the GC heap. Here is an example:
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+
+Found 3 roots.
+```
+Now you need to figure out where the root is located so that you can fix it. The easiest case is when the root is a stack slot or a processor register. In that case, the `gcroot` shows you the name of the function whose frame contains the root and the thread executing that function. The difficult case is when the root is a static variable or a GC handle. 
````
```suggestion
Now you need to figure out where the root is located so you can fix it. The easiest case is when the root is a stack slot or a processor register. In that case, the `gcroot` shows you the name of the function whose frame contains the root and the thread executing that function. The difficult case is when the root is a static variable or a GC handle. 
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```
+Now you need to figure out where the root is located so that you can fix it. The easiest case is when the root is a stack slot or a processor register. In that case, the `gcroot` shows you the name of the function whose frame contains the root and the thread executing that function. The difficult case is when the root is a static variable or a GC handle. 
+
+In our example above, the first root is a local of type `System.Reflection.RuntimeMethodInfo` stored in the frame of the function `example.Program.Main(System.String[])` at address `rbp-20` (`rbp` is the processor register `rbp` and -20 is a hexadecimal offset from that register).
````
```suggestion
In the previous example, the first root is a local of type `System.Reflection.RuntimeMethodInfo` stored in the frame of the function `example.Program.Main(System.String[])` at address `rbp-20` (`rbp` is the processor register `rbp` and -20 is a hexadecimal offset from that register).
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+## Debugging unloading issues
+Debugging issues with unloading can be tedious. You can get into situations where you don't know what can be holding an `AssemblyLoadContext` alive, but the unload fails.
+The best weapon to help with that is Windbg (LLDB on Unix) with the SOS plugin. We need to find what's keeping a `LoaderAllocator` belonging to the specific `AssemblyLoadContext` alive.
+This plugin allows us to look at GC heap objects, their hierarchies and roots. 
```
```suggestion
This plugin allows you to look at GC heap objects, their hierarchies, and roots.
```

mairaw commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+```console
+!gcroot -all 0x000002b78000ce40 
+```
+This will dump the chain of object references that lead to the `LoaderAllocator` instance. The list starts with the root, which is the entity that keeps our `LoaderAllocator` alive and thus is the core of the problem we are debugging. The root can be a stack slot, a processor register, a GC handle, or a static variable.
````
```suggestion
This command dumps the chain of object references that lead to the `LoaderAllocator` instance. The list starts with the root, which is the entity that keeps our `LoaderAllocator` alive and thus is the core of the problem you're debugging. The root can be a stack slot, a processor register, a GC handle, or a static variable.
```

mairaw reviewed (changes requested):
Thanks for doing this @janvorli. I've added some initial comments to get this going.

mairaw commented:
Sounds good @rpetrusha about the link to the API!

janvorli commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+## Example source with unloadability issues
+This example is used in the debugging above.
+### Main testing program
+```csharp
````
I have a pending PR to put all the snippets into that repo. After it is approved and merged, I'll update this doc to not to have any inline code.

janvorli commented:
@mairaw thank you so much for the review and the info on the docs authoring pack. I didn't know such a thing exists.

janvorli commented on docs/core/tutorials/unloadability-howto.md:
````diff
@@ -0,0 +1,362 @@
+        return null;
+    }
+}
+```
````
Ah, got it, after installing the doc authoring pack, it complains that there should be an empty line after that.

janvorli commented on docs/core/tutorials/unloadability-howto.md:
```diff
@@ -0,0 +1,362 @@
+There is one noteworthy difference between the unloading using `AssemblyLoadContext` and using AppDomains. With AppDomains, the unloading is forced. At the unload time, all threads running in the target AppDomain are aborted, managed COM objects created in the target AppDomain are destroyed, etc. With `AssemblyLoadContext`, the unload is "cooperative". Calling the <xref:System.Runtime.Loader.Asse ...
+- No threads have code from the assemblies loaded into the `AssemblyLoadContext` on their call stacks.
+- There are no strong references to types from the assemblies loaded into the `AssemblyLoadContext`, instances of those types and the assemblies themselves. The GC needs to collect those first.
+## Using collectible AssemblyLoadContext
```
I am not sure what would you like to see here. You can use Visual Studio or just the SDK package with any editor you want. This seems general to any .NET Core development.

As for the section on debugging issues, it already mentions that you'll need WinDbg on Windows or LLDB on Unix.

mairaw replied:
Yep, but any tutorial that we do we list what are the prereqs.

Here are some examples:
https://docs.microsoft.com/en-us/dotnet/core/tutorials/with-visual-studio-code#prerequisites
https://docs.microsoft.com/en-us/dotnet/core/tutorials/with-visual-studio#prerequisites
https://docs.microsoft.com/en-us/dotnet/core/tutorials/using-with-xplat-cli#prerequisites

janvorli commented:
@rpetrusha, regarding the broken link - I wonder if the link is really correct, as it contains `%2A` in the middle. Or does it have a special meaning?
`<xref:System.Runtime.Loader.AssemblyLoadContext.Unload%2A?displayProperty=nameWithType>`

rpetrusha commented:
@janvorli, ordinarily, links have to be fully qualified with parameter types. %2A indicates that the link should be to the overload page rather than to a specific overload.

rpetrusha commented:
Closing and reopening to begin new build after dotnet/samples#617 was merged.

janvorli commented:
@rpetrusha, @mairaw  I've updated the doc so that the last two code blocks are not inline anymore and point to the samples repo. As for the rest of those, I am actually wondering whether to leave them all inline or whether to update the doc to replace the ones that are compilable with a reference to the samples and only keep the snippets that cannot be compiled on their own inline.
Do you have a preference w.r.t. that?

janvorli commented:
@rpetrusha, @mairaw can you please reply to my previous comment? I would like to merge the doc in soon.

rpetrusha commented:
I apologize for the long delay in responding, @janvorli. The best option is to place all compilable samples in the samples repo, with small snippets inline.  In addition, though, some of the inline snippets can be referenced from the samples repo. For example, the currently inline code 
```
for (int i = 0; testAlcWeakRef.IsAlive && (i < 10); i++)
{
    GC.Collect();
    GC.WaitForPendingFinalizers();
}
```
is included in your main testing program. So we'd want to pull that snippet into the build, instead of having two versions existing independently. The way to do that is to add a sniippet tag with a label (the snippet tags are removed from the documentation when it's built):
```
// <Snippet1>
for (int i = 0; testAlcWeakRef.IsAlive && (i < 10); i++)
{
    GC.Collect();
    GC.WaitForPendingFinalizers();
}
// </Snippet1>
```
Then add the reference to the snippet to the topic by including the snippet label without the string "snippet":

\[!code-csharp[Main testing program]\(~/samples/snippets/core/tutorials/unloading/unloadability_issues_example_main.cs#1)]

But it's OK to leave most the remaining one- or two-line snippets in the "Use a custom collectible AssemblyLoadContext" section inline, since they don't exactly correspond to code that would go in the repo.

mairaw commented:
Also closing and reopening this to see if the broken link would go away - it should since I loaded the 3.0 APIs.

rpetrusha commented:
Thanks, @mairaw; I'd meant to close and reopen, but I forgot. The broken link should definitely go away.

There are a couple of additional things to do, @janvorli:
- Maira and I thought that the best location for this topic would be a how-to directory (dotnet/docs/core/how-to) rather than the tutorials directory. Could you create the directory and move the topic to it? (That also means that you should change the path in the samples-and-tutorials/index.md document in your PR.)
- That also requires the creation of the an index.md page whose title is ".NET Core how-to articles"  for the directory, which at this point would just have a single entry (your topic).
- It also requires that both the top-level how-to-page and your article be added to the TOC (found in dotnet/docs/docs/toc.md).
- Finally, it would be good to add a link to your article to the managed reference documentation. Once your PR is merged, I can take care of doing that.

janvorli commented:
@rpetrusha, @mairaw  thank you for your feedback. Could you please give me an advise on the description  / title in the new `dotnet/docs/core/how-to/index.md`? I have hard time coming up with wording that would differentiate it from the tutorials and make it general enough for possible future articles in this area.

janvorli commented:
And I have additional questions:
Can snippets be embedded in other snippets? For example, can I do this?
```csharp
class Test
{    
    // <Snippet2>
    [MethodImpl(MethodImplOptions.NoInlining)]
    static int ExecuteAndUnload(string assemblyPath, out WeakReference alcWeakRef)
    {
    // <Snippet3>
        var alc = new TestAssemblyLoadContext();
        Assembly a = alc.LoadFromAssemblyPath(assemblyPath);
    // </Snippet3>

        alcWeakRef = new WeakReference(alc, trackResurrection: true);

    // <Snippet4>
        var args = new object[1] { new string[] {"Hello"}};
        int result = (int)a.EntryPoint.Invoke(null, args);
    // </Snippet4>

    // </Snippet5>
        alc.Unload();
    // </Snippet5>

        return result;
    }
    // </Snippet2>
}
```

And when you say the sample should be compilable, what does it really mean? Should it contain all the "using xxxx" and have methods placed inside of classes so that it can be pasted into a .cs file and compiled without any additional touches?

rpetrusha commented:
Yes, @janvorli,  snippets can be embedded in other snippets.
"Compilable" means that the source code in dotnet/samples should be able to be compiled from the command line (for .NET Framework source) or with the dotnet CLI. We've also usually been including a *.csproj or *.vbproj file for source code that targets .NET Core. You can choose whether to present a all of the source code file in the docs (in that case, it should be compilable, with using statements, etc., if it is copied and pasted) or just a code fragment.

rpetrusha commented:
I also suspect that I gave you some bad guidance, @janvorli. In thinking about it, rather than creating a separate .NET Core how to node in the TOC, it's better to locate this topic in the Assemblies in .NET section of the .NET Guide. The section is sparse (it only has a top-level overview topic and a child topic on the assembly file format) largely because a while ago assembly-related topics were treated as language topics and duplicated in the C# and Visual Basic documentation. We're planning to consolidate the documentation by moving it to the .NET Guide.

So the file can be moved to ~/docs/docs/standard/assembly/. Then the index.md file in that directory can be updated and the article can be added to the TOC at ~/docs/docs/toc.md. 

I'm also wondering if we should rename it to "How to: Load, unload, and debug assembly loading in .NET Core".

janvorli commented:
@rpetrusha I've updated the snippets in the samples repo so that they use the <snippetXX> tags and  thus reduced their number. I've also moved them to a new location that corresponds to the new suggested location of the doc. Can you please take a look so that I can update my doc accordingly?

rpetrusha commented:
I've merged dotnet/samples#677, @janvorli.

janvorli commented:
I've somehow screwed the commits. Will fix it soon.

janvorli commented:
@karelz, @JRAlexander, @BillWagner, @cartermp, @Lxiamail - please ignore the request to review. I have not sent them, they somehow got added automatically based on a bad merge commit that I've fixed.

janvorli commented:
@rpetrusha I've tried to reference the snippets, but it works only for the snippet #1, for the rest it complains that the links are invalid. I have no idea what could be causing that. Do you have any idea what can be wrong?

mairaw commented on docs/standard/assembly/unloadability-howto.md:
```diff
@@ -0,0 +1,247 @@
+
+You can create an instance of the custom `AssemblyLoadContext` and load an assembly into it as follows:
+
+!code-csharp[Part 1](~/samples/snippets/standard/assembly/unloading/simple_example.cs#3)]
```
I think you're just missing part of the syntax 

```suggestion
[!code-csharp[Part 1](~/samples/snippets/standard/assembly/unloading/simple_example.cs#3)]
```

janvorli replied:
Ah, it was right under my nose! Thank you!

mairaw commented:
Also, feel free to just push new commits, since we squash all of them when we merge. We like to have each PR as a single commit in our history.

mairaw commented on docs/standard/assembly/unloadability-howto.md:
````diff
@@ -0,0 +1,247 @@
+```
+
+The command means "apply to all threads the `!clrstack` command". Here is the output of that command for our example. Unfortunately, LLDB on Unix doesn't have any way to apply a command to all threads, so you'll need to resort to manually switching threads and repeating the `clrstack` command.
+You should ignore all threads where the debugger says "Unable to walk the managed stack."
````
```suggestion
Ignore all threads where the debugger says "Unable to walk the managed stack."
```

mairaw commented on docs/standard/assembly/unloadability-howto.md:
````diff
@@ -0,0 +1,247 @@
+~*e !clrstack
+```
+
+The command means "apply to all threads the `!clrstack` command". Here is the output of that command for our example. Unfortunately, LLDB on Unix doesn't have any way to apply a command to all threads, so you'll need to resort to manually switching threads and repeating the `clrstack` command.
````
```suggestion
The command means "apply to all threads the `!clrstack` command". The following is the output of that command for the example. Unfortunately, LLDB on Unix doesn't have any way to apply a command to all threads, so you'll need to resort to manually switching threads and repeating the `clrstack` command.
```

mairaw commented on docs/standard/assembly/unloadability-howto.md:
```diff
@@ -0,0 +1,247 @@
+
+## Program loaded into the TestAssemblyLoadContext
+
+This is the `test.dll` passed to the `ExecuteAndUnload` method in the main testing program.
```
```suggestion
The following code represents the `test.dll` passed to the `ExecuteAndUnload` method in the main testing program.
```

mairaw commented on docs/standard/assembly/unloadability-howto.md:
```diff
@@ -0,0 +1,247 @@
+
+## Example source with unloadability issues
+
+This example is used in the debugging above.
```
```suggestion
The following code is used in the previous debugging example.
```

mairaw reviewed (approved):
I'm approving the PR but left a few suggestions for you to consider. @rpetrusha will have to approve it as well since he requested some changes earlier.

mairaw committed:
Update docs/standard/assembly/unloadability-howto.md

Co-Authored-By: janvorli <janvorli@microsoft.com>

mairaw committed:
Update docs/standard/assembly/unloadability-howto.md

Co-Authored-By: janvorli <janvorli@microsoft.com>

mairaw committed:
Update docs/standard/assembly/unloadability-howto.md

Co-Authored-By: janvorli <janvorli@microsoft.com>

mairaw committed:
Update docs/standard/assembly/unloadability-howto.md

Co-Authored-By: janvorli <janvorli@microsoft.com>

janvorli commented:
Closing and reopeing so that it picks a typo fix in a snippet tag in the samples repo

rpetrusha commented:
This looks good, @janvorli. I'll merge your PR now.
